core/store/src/trie/shard_tries.rs

```diff
@@ -210,12 +216,27 @@ impl ShardTries {
     }
 
     pub fn update_cache(&self, ops: Vec<(&CryptoHash, Option<&[u8]>)>, shard_uid: ShardUId) {
-        let mut caches = self.0.caches.write().expect(POISONED_LOCK_ERR);
-        let cache = caches
-            .entry(shard_uid)
-            .or_insert_with(|| TrieCache::new(&self.0.trie_config, shard_uid, false))
-            .clone();
-        cache.update_cache(ops);
+        // First acquire a read lock to see if the shard uid exists. The shard uid may not exist due to resharding
+        let shard_uid_exists = {
+            let caches = self.0.caches.read().expect(POISONED_LOCK_ERR);
+            caches.contains_key(&shard_uid)
+        };
+        // If the shard uid exists, we acquire a read lock to update the trie cache for the corresponding shard.
+        // If the shard uid does not exist, then we acquire a write lock to expand the cache with the new shard uid as a key.
+        if shard_uid_exists {
+            let caches = self.0.caches.read().expect(POISONED_LOCK_ERR);
+            match caches.get(&shard_uid) {
+                Some(cache) => cache.update_cache(ops),
+                None => debug_assert!(false, "key existence has been checked"),
+            }
+        } else {
+            let mut caches = self.0.caches.write().expect(POISONED_LOCK_ERR);
+            let cache = caches
+                .entry(shard_uid)
+                .or_insert_with(|| TrieCache::new(&self.0.trie_config, shard_uid, false))
+                .clone();
+            cache.update_cache(ops);
+        }
```

Comment:
@nagisa is there a way to avoid acquiring the read lock twice or copying `ops`?

Reply:
@nagisa Sounds good. I'll try the Mutex approach. But more generally, what do you think of switching to `parking_lot`? We didn't do it back in the day because parking lot does not have lock poisoning detection iirc, but a few years have passed since then and I imagine things have changed...

Reply:
I think the motivation to move towards `parking_lot` was stronger before than it is now. `libstd` has incorporated some of the lessons from parking_lot over the years and nowadays one of the stronger motivations would be the access to additional APIs not available in libstd.

poisoning on the other hand is really contentious with regards to how useful it is even. T-libs did some surveys about whether people are handing it at all in any way or form and rough consensus is that lock poisoning is largely a complexity nobody is willing to deal with. For us with `panic=abort` that is doubly true -- the locks can't possibly ever get poisoned, but even if they do, we don't handle that eventuality anywhere, nor do we actually make data structure modifications such that internal invariants would be invalidated most of the time.